### PR TITLE
Sovereign epoch start code placeholder

### PR DESCRIPTION
**Pull request by mariusmihaic**
## Reasoning behind the pull request
- First sovereign epoch change code placeholders
  
## Proposed changes
- Epoch start trigger metachain used for sovereign chain block - also created factories
- Refactored sovereign block processor and factories
- We need another sov chain simulator (which embeds chain sim) struct which will have different private funcs like checking epoch change - TBD how we leave this in the end, can stay as it is for now

## Testing procedure
- 
- 
- 

## Pre-requisites

Based on the [Contributing Guidelines](https://github.com/multiversx/mx-chain-go/blob/master/.github/CONTRIBUTING.md#branches-management) the PR author and the reviewers must check the following requirements are met:
- was the PR targeted to the correct branch?
- if this is a larger feature that probably needs more than one PR, is there a `feat` branch created?
- if this is a `feat` branch merging, do all satellite projects have a proper tag inside `go.mod`?

### CODE REVIEW

File: factory/processing/processComponents_test.go [hunk 1]
```diff
@@ -999,6 +999,7 @@ func getRunTypeComponents(rt runType.RunTypeComponentsHolder) *mainFactoryMocks.
 		ExtraHeaderSigVerifier:              rt.ExtraHeaderSigVerifierHolder(),
 		GenesisBlockFactory:                 rt.GenesisBlockCreatorFactory(),
 		GenesisMetaBlockChecker:             rt.GenesisMetaBlockCheckerCreator(),
+		EpochStartTriggerFactoryField:       rt.EpochStartTriggerFactory(),
```
Comment (axenteoctavian) [inline factory/processing/processComponents_test.go:1002]: In this file at line 950 you need to add one more test for `nil EpochStartTrigger`

File: factory/runType/runTypeComponentsHandler.go
Comment (axenteoctavian) [inline factory/runType/runTypeComponentsHandler.go:1]: Missing EpochStartTriggerFactory nil check at line 200.

File: node/chainSimulator/sovereignChainSimulator.go [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+		}
+
+		if epochReachedOnAllNodes {
+			return nil
+		}
+	}
+	return fmt.Errorf("exceeded rounds to generate blocks")
+}
+
+func (ss *sovereignChainSimulator) isSovereignTargetEpochReached(targetEpoch int32) (bool, error) {
```
Comment (axenteoctavian) [inline node/chainSimulator/sovereignChainSimulator.go:46]: You could also refactor the existing func `isTargetEpochReached` and not change the sovereign chain simulator.
But this approach is fine anyway, meaning we won't change the normal chain simulator files and maybe later we'll want different behavior for other funcs.

File: integrationTests/testProcessorNode.go [hunk 3]
```diff
@@ -2267,182 +2267,10 @@ func (tpn *TestProcessorNode) initBlockProcessor() {
...
-			ArgBaseProcessor:             argumentsBase,
-			SCToProtocol:                 scToProtocolInstance,
-			PendingMiniBlocksHandler:     &mock.PendingMiniBlocksHandlerStub{},
-			EpochEconomics:               epochEconomics,
-			EpochStartDataCreator:        epochStartDataCreator,
-			EpochRewardsCreator:          epochStartRewards,
-			EpochValidatorInfoCreator:    epochStartValidatorInfo,
-			ValidatorStatisticsProcessor: tpn.ValidatorStatisticsProcessor,
-			EpochSystemSCProcessor:       epochStartSystemSCProcessor,
-		}
```
Comment (axenteoctavian) [inline integrationTests/testProcessorNode.go:2272]: You can delete this empty line

File: factory/epochStartTrigger/epochStartTriggerFactory.go
Comment (axenteoctavian) [inline factory/epochStartTrigger/epochStartTriggerFactory.go:1]: `_test` file missing

File: factory/epochStartTrigger/sovereignEpochStartTriggerFactory.go
Comment (axenteoctavian) [inline factory/epochStartTrigger/sovereignEpochStartTriggerFactory.go:1]: `_test` file missing

File: process/block/metablock.go [hunk 4]
```diff
@@ -786,7 +761,7 @@ func (mp *metaProcessor) CreateBlock(
 	return metaHdr, body, nil
 }
 
-func (mp *metaProcessor) isPreviousBlockEpochStart() (uint32, bool) {
+func (mp *baseProcessor) isPreviousBlockEpochStart() (uint32, bool) {
```
Comment (axenteoctavian) [inline process/block/metablock.go:764]: Please move this func in `process/block/baseProcess.go`

File: node/chainSimulator/sovereignChainSimulator.go (hunk 2 above)
Comment (sasurobert) [inline node/chainSimulator/sovereignChainSimulator.go:46]: error is not returned here. will it be changed ? every return is nil.

File: process/block/sovereignChainBlock.go [hunk 5]
```diff
@@ -1138,6 +1276,25 @@ func (scbp *sovereignChainBlockProcessor) CommitBlock(headerHandler data.HeaderH
 	return nil
 }
 
+func (scbp *sovereignChainBlockProcessor) commitEpochStart(header data.HeaderHandler, body *block.Body) {
+	if header.IsStartOfEpochBlock() {
+		scbp.epochStartTrigger.SetProcessed(header, body)
+		scbp.epochStartTrigger.SetProcessed(header, body)
```
Comment (axenteoctavian) [inline process/block/sovereignChainBlock.go:1282]: Dupplicate

File: testscommon/mainFactoryMocks/runTypeComponentStub.go [hunk 6]
```diff
@@ -109,6 +112,7 @@ func NewRunTypeComponentsStub() *RunTypeComponentsStub {
 		ExtraHeaderSigVerifier:              &headerSigVerifier.ExtraHeaderSigVerifierHolderMock{},
 		GenesisBlockFactory:                 &testFactory.GenesisBlockCreatorFactoryMock{},
 		GenesisMetaBlockChecker:             &testFactory.GenesisMetaBlockCheckerMock{},
+		EpochStartTriggerFactoryField:       epochStartTrigger.NewEpochStartTriggerFactory(),
```
Comment (axenteoctavian) [inline testscommon/mainFactoryMocks/runTypeComponentStub.go:115]: Maybe use mock here if it's possible.

File: process/block/metablock.go (hunk 4 above)
Comment (sasurobert) [inline process/block/metablock.go:764]: move this to baseprocessor if you renamed the pointer structure.

File: factory/processing/processComponents_test.go (hunk 1 above)
Comment (mariusmihaic) [reply factory/processing/processComponents_test.go:1002]: Right, good catch, added it

File: factory/runType/runTypeComponentsHandler.go
Comment (mariusmihaic) [reply factory/runType/runTypeComponentsHandler.go:1]: Right, very good comment, added it

File: integrationTests/testProcessorNode.go (hunk 3 above)
Comment (mariusmihaic) [reply integrationTests/testProcessorNode.go:2272]: Right, deleted it

File: factory/epochStartTrigger/epochStartTriggerFactory.go
Comment (mariusmihaic) [reply factory/epochStartTrigger/epochStartTriggerFactory.go:1]: Right, tried to do them now but ran into the old "cylcic import" issue and I need to change too much code in this PR to make it work. I should do it in a separate task.

File: factory/epochStartTrigger/sovereignEpochStartTriggerFactory.go
Comment (mariusmihaic) [reply factory/epochStartTrigger/sovereignEpochStartTriggerFactory.go:1]: Right, tried to do them now but ran into the old "cylcic import" issue and I need to change too much code in this PR to make it work. I should do it in a separate task.

File: process/block/metablock.go (hunk 4 above)
Comment (mariusmihaic) [reply process/block/metablock.go:764]: Right, moved it

File: process/block/sovereignChainBlock.go (hunk 5 above)
Comment (mariusmihaic) [reply process/block/sovereignChainBlock.go:1282]: Right, fixed in future pr, but deleted here also anyway

File: testscommon/mainFactoryMocks/runTypeComponentStub.go (hunk 6 above)
Comment (mariusmihaic) [reply testscommon/mainFactoryMocks/runTypeComponentStub.go:115]: Added mock

File: process/peer/sovereignChainProcess.go [hunk 7]
```diff
@@ -20,6 +20,7 @@ func NewSovereignChainValidatorStatisticsProcessor(validatorStatistics *validato
 		validatorStatistics,
 	}
 
+	// TODO: MX-15586 Analyse this func and refactor it for sovereign epoch change
```
[the remark this answers is not in the document]
Comment (mariusmihaic) [reply process/peer/sovereignChainProcess.go:23]: I see that we call `scbp.validatorStatisticsProcessor.UpdatePeerState` from sovereignChainBlock.go inside `processSovereignBlockTransactions`

File: node/chainSimulator/sovereignChainSimulator.go (hunk 2 above)
Comment (mariusmihaic) [reply node/chainSimulator/sovereignChainSimulator.go]: Right, fixed it

File: process/block/metablock.go (hunk 4 above)
Comment (mariusmihaic) [reply process/block/metablock.go:764]: Right, missed it, fixed.

**Review by sasurobert — APPROVED** [review]

**Review by axenteoctavian — APPROVED** [review]
